### ISSUE TITLE
Image processing pipelines fails: insufficient memory due to breadth-first execution

### ISSUE THREAD

**Issue by fg91** [PrefectHQ/prefect#4946]
## Description

I am prototyping an image processing pipeline very similar to what has been presented [here](https://www.youtube.com/watch?v=SIvLhp7aLBw&ab_channel=Heartex) by @kvnkho :

My goal is to 1) load images from a bucket, 2) rotate them (as a dummy transformation), and 3) save the rotated images to another bucket:

```python
@task
def get_image_uris():
    return fs.ls(f"{bucket_name}")

@task
def load_image_from_bucket(x):
    with fs.open(x, "rb") as f:
        img = Image.open(f)
        img.load()
        return img, x

@task
def rotate_image(x):
    img, fn = x    
    return img.rotate(90), fn

@task
def save_image_to_bucket(x):
    img, fn = x

    with fs.open(fn.replace("...", "..._rot"), "wb") as f:
        img.save(f, "JPEG")
```

When running the tasks with a `DaskExecutor` on my machine

```
if __name__ == "__main__":
    with Flow(
            'Load images from bucket',
            executor=DaskExecutor()
    ) as flow:
        loaded_images = load_image_from_bucket.map(get_image_uris)
        rotated_images = rotate_image.map(loaded_images)
        save_image_to_bucket.map(rotated_images)

    state = flow.run()
```

the `load_image_from_bucket` task is executed exclusively (besides of course `get_image_uris`):

```
...
[2021-09-07 11:51:21+0200] INFO - prefect.TaskRunner | Task 'load_image_from_bucket[3720]': Starting task run...
[2021-09-07 11:51:21+0200] INFO - prefect.TaskRunner | Task 'load_image_from_bucket[3724]': Finished task run for task with final state: 'Success'
[2021-09-07 11:51:21+0200] INFO - prefect.TaskRunner | Task 'load_image_from_bucket[3743]': Starting task run...
[2021-09-07 11:51:21+0200] INFO - prefect.TaskRunner | Task 'load_image_from_bucket[3746]': Starting task run...
[2021-09-07 11:51:21+0200] INFO - prefect.TaskRunner | Task 'load_image_from_bucket[3744]': Starting task run...
[2021-09-07 11:51:21+0200] INFO - prefect.TaskRunner | Task 'load_image_from_bucket[3745]': Finished task run for task with final state: 'Success'
...
```

As all the downloaded images are kept in memory, the used memory rises, the machine starts to swap until the flow eventually fails without having uploaded a single transformed image (mapping over 50k images):

<img width="555" alt="Screenshot 2021-09-07 at 11 51 29" src="https://user-images.githubusercontent.com/36511035/132333126-c79c8ef2-40d7-4c91-ad31-2bee47c84a79.png">

The same happens when using an ephemeral dask cluster in kubernetes:

<img width="731" alt="Screenshot 2021-09-07 at 12 18 08" src="https://user-images.githubusercontent.com/36511035/132333253-fd8cf7d5-e17c-4128-b7cc-d07e920a904d.png">

Where the memory dropped temporarily, individual dask worker pods where evicted due to memory pressure.

```
[2021-09-07 10:12:01+0000] INFO - prefect.TaskRunner | Task 'rotate_image[469]': Finished task run for task with final state: 'Success'
[2021-09-07 10:12:01+0000] INFO - prefect.TaskRunner | Task 'rotate_image[464]': Starting task run...
distributed.core - INFO - Event loop was unresponsive in Nanny for 3.45s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.nanny - INFO - Worker process 24 was killed by signal 9
distributed.nanny - WARNING - Restarting worker
```

Eventually all worker pods show numerous of the following exception and the flow effectively halts, CPU usage drops.

```
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/distributed/worker.py", line 2944, in execute
    args2, kwargs2 = self._prepare_args_for_execution(ts, args, kwargs)
  File "/opt/conda/lib/python3.8/site-packages/distributed/worker.py", line 3065, in _prepare_args_for_execution
    data[k] = Actor(type(self.actors[k]), self.address, k, self)
KeyError: 'load_image_from_bucket-3310-6aafac2c4af0408bb5d1653404e2d686'
distributed.worker - ERROR - Exception during execution of task rotate_image-5860-8f9b26fb3c31423eb9567ba4cfad2d33.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/distributed/worker.py", line 3061, in _prepare_args_for_execution
    data[k] = self.data[k]
  File "/opt/conda/lib/python3.8/site-packages/zict/buffer.py", line 80, in __getitem__
    raise KeyError(key)
KeyError: 'load_image_from_bucket-5860-4f390ad7f51348b2b0d910a4115d0cb2'

During handling of the above exception, another exception occurred:
```

## Expected Behavior

What I would want to happen is depth first execution:
Download a number of images, transform, and upload them, then do the same for the next batch.

I am aware of [this PR introducing depth-first execution](https://github.com/PrefectHQ/prefect/pull/2646) and [this blog post explaining it](https://medium.com/the-prefect-blog/map-faster-mapping-improvements-in-prefect-0-12-0-7cacc3f14e16) (hence the people I tagged here @cicdw , @jcrist, @joshmeek, @jlowin, @lauralorenz ).
Running the minimal working example from this PR (that showcases depth-first execution using tasks that take a different amount of time to complete) works as expected for me.

(Sidenote I'm seeing this warning that the blog post mentions to be fixed:
```
//anaconda3/envs/sandbox/lib/python3.8/site-packages/distributed/worker.py:3862: UserWarning: Large object of size 1.65 MiB detected in task graph:
  {'task': <Task: load_image_from_bucket>, 'state':  ... _parent': True}
Consider scattering large objects ahead of time
with client.scatter to reduce scheduler burden and
keep data on workers

    future = client.submit(func, big_data)    # bad

    big_future = client.scatter(big_data)     # good
    future = client.submit(func, big_future)  # good
  warnings.warn(
Handling connection for 63879
```
)

So my question is whether there is a way to enforce or encourage DFE (when the tasks don't take different amounts of time)?

I am also aware of these resources but currently don't know how to use the suggestions to solve the problem:
1. [this closed issue](https://github.com/PrefectHQ/prefect/issues/4556) where the author @peterroelants states that he solved this problem using [dependet flows](https://docs.prefect.io/core/idioms/flow-to-flow.html)
2. [this discussion](https://github.com/PrefectHQ/prefect/discussions/3491) by @snenkov.

I'm happy to work on this given some guidance if a PR in prefect is required to fix this.

## Reproduction
1. A minimal working example that mocks the behaviour:

```python
from prefect import Flow, task

@task
def get_imgs_uris():
    return [i for i in range(10)]

@task
def load_image_from_bucket(x):
    print(f"load image {x}")
    return x

@task
def rotate_image(x):
    print(f"rotate img {x}")
    return x

@task
def save_image_to_bucket(x):
    print(f"save img {x}")
    return


if __name__ == "__main__":

    import prefect
    from prefect.executors import LocalDaskExecutor, DaskExecutor

    with Flow("mapping-test") as flow:
        save_image_to_bucket.map(rotate_image.map(load_image_from_bucket.map(get_imgs_uris)))

    #flow.run(executor=LocalDaskExecutor())
    flow.run(executor=DaskExecutor())
```

2. (almost) working example really downloading and transforming images from a bucket. You would only have to enter the name of a bucket (in the two lines I marked with a #) containing sufficient images to create a wide enough flow to create memory pressure (I tried with 50k).

```python
import fsspec
from fsspec.core import get_fs_token_paths
import io
from PIL import Image

import prefect
from prefect import Flow, task
from prefect.executors import LocalDaskExecutor, DaskExecutor
from prefect.run_configs import KubernetesRun
from dask_kubernetes import KubeCluster, make_pod_spec
from prefect.storage import Docker

bucket_name = "..."                                                                                                         #
bucket_url= f'gs://{bucket_name}'

fs = get_fs_token_paths(bucket_url)[0]

@task
def get_image_uris():
    return fs.ls(f"{bucket_name}/cats_full")

@task
def load_image_from_bucket(x):
    with fs.open(x, "rb") as f:
        img = Image.open(f)
        img.load()
        return img, x

@task
def rotate_image(x):
    img, fn = x    
    return img.rotate(90), fn

@task
def save_image_to_bucket(x):
    img, fn = x

    with fs.open(fn.replace("...", "..._rot"), "wb") as f:                                                                      # 
        img.save(f, "JPEG")

        
if __name__ == "__main__":
    with Flow(
            'Load images from bucket',
            # run_config=KubernetesRun(),
            # executor=DaskExecutor(
            #     cluster_class=lambda: KubeCluster(
            #         make_pod_spec('daskdev/dask:latest'),
            #         namespace="default",
            #         env={'EXTRA_PIP_PACKAGES': "prefect gcsfs fsspec"}
            #    ),
            # adapt_kwargs={"minimum": 5, "maximum": 15},
            # ),
    ) as flow:
        loaded_images = load_image_from_bucket.map(get_image_uris)
        rotated_images = rotate_image.map(loaded_images)
        save_image_to_bucket.map(rotated_images)

    state = flow.run()
```

## Environment
```
{
  "config_overrides": {},
  "env_vars": [],
  "system_information": {
    "platform": "macOS-10.14.5-x86_64-i386-64bit",
    "prefect_backend": "cloud",
    "prefect_version": "0.15.4",
    "python_version": "3.8.11"
  }
}
```

**Comment by madkinsz** [comment]
Hey @fg91 -- thanks for the thorough issue. In the future though, please don't ping a bunch of our team directly.

DFE will not solve your problem here, as all of the task results are held in the task states until the entire flow completes. Subflows help with this (as you noted in a linked issue) because each subflow will release its task states on completion. Generally, we don't recommend passing large data objects between tasks like this as it will cause memory contention. Dask also does not recommend splitting data operations into separate tasks.

I appreciate your offer to PR a fix! I'm happy to help guide you if we find a concrete path to helping your use-case, but some parts of this (like flows holding on to all task states) are pretty deep rooted.

**Comment by jacobdanovitch** [comment]
I have an extremely similar use case where almost all of our data is in raw multi-line text files when it comes in and we more or less just wrap it in a JSON object, maybe apply one intermediate function, and then save it. So same `t1.map(t2.map(t3.map(x)))` workflow but with much smaller items than images and I'm still getting hit by this too. I've had to more or less make everything an `@dask.delayed` and then write all of my logic in a moderately ugly task that calls `client.compute` (though this still doesn't work well, fwiw). 

I think the common theme here is that it's convenient/easy to understand/good practice to split our logic into many smaller tasks (as is encouraged by the docs), but sometimes we have a sequence of a few tasks strung together where we don't really care about the inputs and outputs of the intermediate steps from the lens of observing and persisting their states, since it's really one piece of logic we've divided into several smaller steps. What about something like task composition as a hint for when to treat a sequence this way? For example, something like:

```python
with Flow(...) as flow:
        uris = get_image_uris()
        with compose():
            loaded_images = load_image_from_bucket.map(uris)
            rotated_images = rotate_image.map(loaded_images)
            save_image_to_bucket.map(rotated_images)
```

We could have some context manager (or whatever's easiest, doesn't matter) to indicate that `load_image_from_bucket`, `rotate_image`, and `save_image_to_bucket` should be treated as a single task, combining the inputs and outputs and state and so on. Would this reduce memory usage enough to be worthwhile? Obviously we could just rewrite them in one big function and map it over `uris` to accomplish the same logic but the above feels more idiomatic of Prefect.
